Add plugins for Arris Cable Modem SB6183

utopiabound opened the pull request:
This provides two plugins:
* arris-sb6183 - Power levels, SNR, error counts
* arris-sb6183_uptime - Uptime in days

These are two seperate plugins because it's two different
urls it needs to hit, and this seemed simpler.

arris-sb6183 provides multigraphs for each channel.

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/python
```
I think, you are using python3 syntax here. Thus please clarify this by using `python3`. Additionally it is recommended to allow local overrides of the specific python interpreter (including a virtualenv) by using `#!/usr/bin/env python3` instead.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/python
+
+# modem:
```
Please move these comments into the perldoc documentation below.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+import os
+import sys
+import requests
+from lxml import html
```
I would suggest to use a standard library instead (e.g. https://docs.python.org/3/library/xml.dom.html). Otherwise you force users to install an additional package and you will need to handle the potentially missing library in the code below (by failing with a reasonable error message in case of an `ImportError`).

utopiabound replied:
I would have liked to, but the HTML from the router is poorly structured and xml.etree throws up on parsing it.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+from lxml import html
+
+
+URL = os.getenv("url", "http://192.168.100.1/RgConnect.asp")
```
I would suggest to reduce the configurable part of the URL to the scheme and the host (omitting the path). This will allow you to adjust to a changed URL provided by a later version of a firmware within the plugin code (without the user being forced to change his configuration). For example: `http://192.168.100.1`.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+env.hostname modem
+
+=head1 VERSION
+
```
Please mention the other configuration settings (`url`, `up`, `down`) here, too.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+
+URL = os.getenv("url", "http://192.168.100.1/RgConnect.asp")
+HOSTNAME = os.getenv("hostname", None)
+UPCOUNT = int(os.getenv("up", 4))
```
What is the purpose of this `up` variable? Is it related to slots/ports of the physical device? I would suggest to use a self-explaining variable name here (especially for the configuration variable).

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+        for i in range(1, UPCOUNT + 1):
+            print("up_{0}.label Up Ch {1}".format(i, i))
+            print("up_{0}.type GAUGE".format(i))
+            # print("up_{0}.draw LINE1".format(i))
```
Please remove commented code.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+import re
+import os
+import sys
+import requests
```
Maybe `urllib` (the standard library) would be sufficient here? Fewer dependencies and less code for handling the potentially missing library.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+    if sys.argv[1] == "autoconfig":
+        try:
+            page = requests.get(URL)
+        except:
```
Please avoid bare exceptions (see [PEP8](https://www.python.org/dev/peps/pep-0008/)). I guess, `IOError` would be suitable here.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+        sys.exit(0)
+
+    if sys.argv[1] == "autoconfig":
+        try:
```
Here you would also need to verify the availability of the non-core modules (`lxml` and `requests`).

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,267 @@
+headings = ["".join(x.itertext()).strip() for x in trs.pop(0).findall("td")]
+# ['Channel', 'Lock Status', 'Modulation', 'Channel ID', 'Frequency', 'Power', 'SNR', 'Corrected', 'Uncorrectables']
+
+mapper = lambda x, y: (x, y)
```
I guess, `zip` would do the same, or? (in a list context)

sumpfralle reviewed (commented):
Thank you for contributing these plugins!

Please take a look at my notes below.

Regarding the separate plugins: I think, it would be very beneficial to merge both plugins into one (since you are using `multigraph` already). The separate URLs (or *paths*) should be no issue, since it would be advisable to configure the URL without the path, anyway (see one of my comments below). This would reduce the burden of maintaining two copies of similar code.

Maybe you also want to add [examples graphs](http://guide.munin-monitoring.org/en/latest/develop/plugins/plugin-gallery.html#rules-for-plugin-contributors)? This would be perfect.

Thank you!

utopiabound commented:
Example Output:
![arris_error-day](https://user-images.githubusercontent.com/2285913/94147203-fc280800-fe42-11ea-9419-79f53909e874.png)
![arris_snr-day](https://user-images.githubusercontent.com/2285913/94147204-fc280800-fe42-11ea-8ac8-9987c869890d.png)
![arris_uptime-day](https://user-images.githubusercontent.com/2285913/94147205-fc280800-fe42-11ea-9fd5-cef94f867e51.png)

I've updated plug-in to be single file, and added more error handling.

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,366 @@
+        rxcomment.sub(
+            "",
+            rxblank.sub(
+                " ", "".join(map(lambda x: x.decode("utf-8"), resp.readlines()))
```
Just a style recommendation - maybe a generator comprehension would be easier to read:
```python
"".join(line.decode() for line in resp.readlines())
```

Or maybe just use `resp.read().decode()`?

sumpfralle commented on plugins/router/arris-sb6183:
```diff
@@ -0,0 +1,366 @@
+from urllib import request
+
+HOSTNAME = os.getenv("hostname", None)
+STATUS_URL = "http://192.168.100.1/RgConnect.asp"
```
Are you sure, that the IP address of the router is always the same? Or maybe make it configurable?

(I leave this up to you, since you know that device ...)

utopiabound replied:
The documentation lists this as default and every router I've had from Arris has had this IP.  (There is no way *I* can change the IP, maybe my cable company can)

niclan replied:
The 192.168.100.1 seems to be industry standard.

sumpfralle replied:
A local IP that cannot be changed by the local administrator/user. That sounds horrible :(

Anyway: in this case it is not necessary to make it configurable.

sumpfralle reviewed (commented):
Thank you for merging these plugins into one!

I added two more minor comments below. But the plugin is ready to be merged, if you prefer that.

Regarding the example graphs: please add them to the sub directory `example-graphs` (following the naming convention described in the [munin plugin documentation](http://guide.munin-monitoring.org/en/latest/develop/plugins/plugin-gallery.html#rules-for-plugin-contributors)).

Thanks!

sumpfralle commented:
Thank you for the new plugin - I just merged it now.

Feel free to add the example graphs in a separate pull request, if you feel like ...
